src/vec_poly.h

```diff
@@ -0,0 +1,234 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __NTTEC_VEC_POLY_H__
+#define __NTTEC_VEC_POLY_H__
+
+#include "gf_base.h"
+#include "vec_vector.h"
+
+namespace nttec {
+namespace vec {
+
+/** A vector of size \f$n\f$ represents a polynomial \f$P(X)\f$ of degree
+ * \f$(n-1)\f$
+ *
+ * Vector: \f$v[0], v[1], ..., v[n-1]\f$ represents the polynomial:
+ *  \f$P(X):= v[0] + v[1] X + ... + v[n-1] X^n\f$
+ */
+template <typename T>
+class Poly : public Vector<T> {
+  public:
+    Poly(gf::Field<T>* field, int n);
+    Poly(Poly<T>* a);
+    int get_deg();
+    T get(int exponent);
+    void set(int exponent, T coef);
+    void derivative();
+    T eval(T x);
+    void mul(Poly<T>* b, int deg_out);
+    void mul_to_x_plus_coef(T coef);
+    void neg();
+    void zero();
+    void dump();
+
+  private:
+    gf::Field<T>* field;
+    T field_characteristic;
+    T* buf;
+    int degree;
+};
+
+template <typename T>
+Poly<T>::Poly(gf::Field<T>* field, int n) : Vector<T>(field, n)
+{
+    this->field = field;
+    this->field_characteristic = field->get_p();
+    this->buf = this->get_mem();
+    this->degree = 0;
+}
+
+template <typename T>
+Poly<T>::Poly(Poly<T>* a) : Vector<T>(a->rn, a->get_n())
+{
+    this->field = (gf::Field<T>*)this->rn;
+    this->buf = this->get_mem();
+    this->field_characteristic = field->get_p();
+
+    std::copy_n(a->get_mem(), this->n, buf);
+    this->degree = a->get_deg();
+}
+
+template <typename T>
+int Poly<T>::get_deg()
+{
+    return degree;
+}
+
+template <typename T>
+T Poly<T>::get(int exponent)
```

Comment:
I think it would be better to overload the method [] instead of providing a method `get`

```
template <typename T>
const T& Poly<T>::operator[] (const int exponent) const
```

But this should be done in another PR, to replace all the `get` by` operator [] `in one sweep.

Maybe we can create an issue tagged improvement if we don't want to tackle it now (so we don't forget).